Add option to reuse existing subvolume

laenion opened the pull request:
If a subvolume exists already mksubvolume will abort. In certain
situations however, e.g. if there was a rollback to a previous snapshot,
it may be desirable to be able to reuse this subvolume.

This will make it possible to use the mksubvolume command in package
scripts.

To avoid having to duplicate the exception handler code multiple times
a C wrapper object was introduced to automatically clean up in the
destructor.

Fixes https://github.com/openSUSE/snapper/issues/236.

aschnell commented on client/mksubvolume.cc:
```diff
@@ -258,6 +267,33 @@ find_filesystem(MntTable& mnt_table)
+struct TmpMountpoint {
+    unique_ptr<char[]> mountpoint;
+    const libmnt_fs* fs;
+    TmpMountpoint(const string tmpMountpoint, const libmnt_fs* libmntfs, const string& subvol_opts)
```
tmpMountpoint should be passed by reference.

laenion replied:
Done.

aschnell commented on client/mksubvolume.cc:
```diff
@@ -332,36 +368,20 @@ doit()
-
-	throw;
+	if (e.error_number == EEXIST)
+	    cout << "subvolume exists already" << endl;
```
What happens if e.g. a directory already exists where the subvolume should be created?

laenion replied:
Not much, as nothing has to be done here - the destructor of TmpMountpoint will be called to clean up everything, and mksubvolume will continue with creating the fstab entry.

aschnell replied:
But adding the fstab entry for anything not a subvolume is wrong. This will lead to boot failures.

laenion replied:
Ah, I see: You mean when it's just a directory instead of a subvolume. At least on openSUSE I hope that will never be the case, as that would mean someone modified the '@' subvolume, but I agree this may happen more easily on other distributions...

laenion replied:
Detecting conflicting files in the latest commit now.

aschnell commented on client/mksubvolume.cc:
```diff
@@ -43,6 +43,7 @@ using namespace BtrfsUtils;
 
 string target;
 
+bool force = false;
```
Where is force used?

laenion replied:
Oh, I'm sorry, this should have been used as part of the if condition in the exception handler above - will fix.

laenion replied:
Thinking about it: It _could_ also just be left out, so existing suvolumes are reused automatically. I'm not sure if this is a good idea though.

laenion commented:
Fixed all remarks. Checking for existing fstab entries is done in the code already, no changes were needed here.

aschnell commented on client/mksubvolume.cc:
```diff
@@ -377,7 +377,14 @@ doit()
     {
 	if (e.error_number == EEXIST && force)
+	{
+	    const string path = string(tmp_mountpoint.mountpoint.get()).append("/").append(subvolume_name);
```
string does have the + operator.

laenion replied:
Changed.

aschnell commented on client/mksubvolume.cc:
```diff
@@ -377,7 +377,14 @@ doit()
+	    const string path = string(tmp_mountpoint.mountpoint.get()).append("/").append(subvolume_name);
+
+	    struct stat sb;
+	    if (stat(path.c_str(), &sb) == 0 && !is_subvolume(sb))
```
Failure of stat is not handled.

laenion replied:
And it should be lstat anyway...

laenion replied:
Error condition reversed to only accept "good" case.

aschnell commented:
Isn't the operation stopped anyway due to the ``` if (access(target.c_str(), F_OK) == 0)```?

laenion commented:
Only if the target mount directory (in contrast to the subvolume itself) exists already. But looking at the rollback case the creation of this directory would also have been rolled back.
However: When initially implementing this patch I indeed had a check in that `if` query to not throw an error when the `-f` option was set. I can re-add that check if desired.

aschnell commented:
AFAIS the target mount directory can be identical to the subvolume if the "root" subvolume is mounted. That was the case with SUSE some time ago.

And rollback could not be the only usecase for the option. So I suggest to disable also that check.

aschnell commented:
If the entry already exists in /etc/fstab the operation also stops.

aschnell commented:
Maybe we could define that the following is a valid use-case: An RPM creates a subvolume in post-install but does not remove the subvolume in post-uninstall. Removing the RPM and installing it again must work.

laenion commented:
Changed the handling according to the (imho reasonable) use-case: An existing target directory will be reused now, and if the target exists in /etc/fstab the mount entry will be removed. If the target it mounted already an attempt to unmount it will be done, but if that fails the operation will still abort.

The new behavior will *only* be triggered with the `-f` option, though RPM `post` scripts should probably always call `mksubvolume` with this option for rollback compatibility...

aschnell commented:
What is the reason for first deleting the fstab entry and unmounting the subvolume and then add and mount again? Why not simply omit the add/mount step in that case?

laenion commented:
My thinking was that there may be artifacts from a previous run which may not match the new entry, but you are correct, this won't happen when not interfering manually. I'll change the code accordingly.

aschnell commented:
There might be other artifacts, e.g. options for the subvolume that are not changed (no-cow is not cleared).

aschnell commented:
What should happen if the user has already mounted a different btrfs filesystem at the target location?
That looks like a valid use-case to me.

Right now the program fails with an exception.

laenion commented:
After some thinking I tend to implement it to just create the missing parts, but with added consistency checks:
* Just leave out the option completely (currently `-f`) and behave reasonably well by default:
  * If the subvolume exists already check if the COW flag matches the command line option
  * If an fstab entry exists already check whether device and subvolume or subvolumeid match the expected new entry
  * If the target is mounted already also check for correct device and subvolume/-id
* If any of those conditions are not true abort with an error.

Advantages:
* Works without any modification to existing packaging scripts
* Exits with an error if the user changed something manually (i.e. the environment doesn't match the expected values which would have been autocreated), creates the missing entries otherwise
* No data loss by forcefully overwriting things
* No inconsistent data by just leaving entries as they are without consistency check

laenion commented:
The latest commits are implementing https://github.com/openSUSE/snapper/pull/488#issuecomment-511906588.

aschnell commented on client/mksubvolume.cc:
```diff
@@ -326,45 +378,79 @@ doit()
+    if (mounted_entry != NULL)
+    {
+        char* subvol_expected;
+	size_t valsz;
```
AFAIS since valsz is not really needed passing NULL to mnt_fs_get_option should be fine.

laenion replied:
Correct, I even did that in the reimplementation of `is_subvol_mount`, but didn't change this occurrence.
Will change.

aschnell commented on client/mksubvolume.cc:
```diff
@@ -326,45 +378,79 @@ doit()
+	    if (strcmp(real_device, mnt_fs_get_source(mounted_entry)) != 0)
+		throw runtime_error("different device mounted on target");
+	    char* subvol_real;
+	    mnt_fs_get_option(mounted_entry, "subvol", &subvol_real, &valsz);
```
Is it OK to not check the return value? What it the subvol entry does not get found?

laenion replied:
This should not have happened, as we knew at this place that the device contains a btrfs file system, which will always have a `subvol` option. Just to be safe I've added the check, though.

aschnell commented on client/mksubvolume.cc:
```diff
@@ -258,6 +277,33 @@ find_filesystem(MntTable& mnt_table)
     }
 }
 
+struct TmpMountpoint {
```
I suggest to make mountpoint and fs private and add a get_path() function.

laenion replied:
Done.

aschnell commented on client/mksubvolume.cc:
```diff
@@ -326,45 +378,79 @@ doit()
+	throw runtime_error("target is a dedicated mountpoint");
+    if (fstab_entry != NULL && strcmp(mnt_fs_get_source(fstab_entry), mnt_fs_get_source(expected_fs)) != 0)
+	throw runtime_error("existing fstab entry doesn't match target device");
+    if (fstab_entry != NULL && strcmp(mnt_fs_get_options(fstab_entry), mnt_fs_get_options(expected_fs)) != 0)
```
Even harmless options like "rw" will make this check fail. Are there not options the user might have added that should not stop the program? Please at least add a comment.

laenion replied:
This was written under the premise that the application will usually be called by scripts with idempotent results from consecutive runs. Manually created or modified entries are deliberately not recognized as "matching", as the result is not the same.

This could be changed, but would significantly increase code complexity (similar to the mounted device detection). Shall I still implement it?

aschnell replied:
Documenting it is also fine for me.

laenion replied:
I've changed it to check only the `subvol` option (so any other option the user may add will be accepted).

As `subvol` may be given either relative or absolute (`@/something` vs. `/@/something`) a normalization will be done before the comparison.

aschnell commented on client/mksubvolume.cc:
```diff
@@ -326,45 +378,79 @@ doit()
+						     MNT_ITER_BACKWARD);
+    if (subvolume_name.empty())
+	throw runtime_error("target is a dedicated mountpoint");
+    if (fstab_entry != NULL && strcmp(mnt_fs_get_source(fstab_entry), mnt_fs_get_source(expected_fs)) != 0)
```
This also fails if the filesystem is mounted e.g. by label in fstab (and / is mounted by something else).

laenion replied:
The problem is that both `fstab_entry` and `expected_fs` are "virtual" entries without any guarantee that they are actually mounted. `real_device` however is determined by looking for the device in the _mtab_ file, because that one will always contain the absolute device path.

I'm not very keen on parsing everything manually, so for the time being I'd opt to not implement this.

laenion replied:
Fixed by using `mnt_resolve_spec`, which does exactly what we want.

aschnell commented on client/mksubvolume.cc:
```diff
@@ -326,45 +378,79 @@ doit()
+	size_t valsz;
+	mnt_fs_get_option(expected_fs, "subvol", &subvol_expected, &valsz);
+	// Map UUID / LABEL to a physical device name
+	const char* real_device = mnt_fs_get_source(mnt_table_find_source(mtab_table,
```
real_device can be null when the entry for / is wrong in fstab (or udev did not generate the by-label link) in which case the program can core dump.

laenion replied:
Fixed.

aschnell commented on client/mksubvolume.cc:
```diff
@@ -326,45 +396,84 @@ doit()
+	if (e.error_number == EEXIST)
 	{
-	}
+	    const string path = string(tmp_mountpoint.get_path()) + "/" + subvolume_name;
```
The string constructor is not needed.

laenion replied:
Fixed.

aschnell commented on client/mksubvolume.cc:
```diff
@@ -259,6 +278,51 @@ find_filesystem(MntTable& mnt_table)
+    }
+
+    string
+    get_path()
```
Should be a const function.

laenion replied:
Fixed.

aschnell commented on client/mksubvolume.cc:
```diff
@@ -326,45 +396,84 @@ doit()
+    char* subvol_expected;
+    libmnt_fs* expected_fs = create_fstab_line(fs, subvol_option, subvolume_name);
+    mnt_fs_get_option(expected_fs, "subvol", &subvol_expected, NULL);
+    assert(subvol_expected != NULL);
```
Please throw an exception.

laenion replied:
Fixed by reworking error handling.

aschnell commented on client/mksubvolume.cc:
```diff
@@ -326,45 +396,84 @@ doit()
+    assert(subvol_expected != NULL);
+
+    // Consistency checks on (partially) existing entries
+    libmnt_fs* fstab_entry = mnt_table.find_target(target, MNT_ITER_FORWARD);
```
Below are two tests for fstab_entry != NULL. Why not throw an exception right here in that case?

aschnell commented:
Have all those pre checks been checked with multiple devices btrfs?

laenion commented:
> Have all those pre checks been checked with multiple devices btrfs?

No, I currently don't have such a setup. I won't make it to set it up before my vacation next week - could I ask you to run these tests?

aschnell commented:
Mutliple devices btrfs does not work correctly. I have a btrfs on sda2 and sda3. In fstab sda3 is used. Anyway in /proc/mounts sda2 is reported. Running the program a second time reports an error ("failure (different device mounted on target)").

BTW: Some debug output when using --verbose could help testing, e.g. the device names detected in this case.

laenion commented:
So in the end it seems a reliable consistency check is not possible, at least I'm not aware of any existing mechanism to get a canonical path for any given source identifier (even unmounted ones)...

I'm really tempted to just go back to the original use case and just ignore an already existing subvolume, without any additional checks or fixes...

laenion commented:
Correction: This may be what `mnt_resolve_spec` does.

laenion commented:
`mnt_resolve_spec` does indeed what I was looking for all the time - it will resolve any source to a canonical path spec. To ensure that the cache - required for this conversion to work - is always initialized I've added the option to set the cache to the MntTable class and changed the code to use it for all mnt_table operations.

lslezak commented on client/mksubvolume.cc:
```diff
@@ -144,20 +149,41 @@ do_add_fstab_and_mount(MntTable& mnt_table, const libmnt_fs* fs, const string& s
+do_add_fstab_and_mount(MntTable& mnt_table, libmnt_fs* x)
+{
+    if (verbose)
+	cout << "do-add-fstab-and-mount" << endl;
```
I know this line was already there, but what's the purpose of this? Is it a debugging message or something for the end user? For debugging it should use `y2deb`/`y2mil`, for end users it should contain a better text and should be translated...

laenion commented:
It seems `mksubvolume` is not using any of the _y2_ helper functions. In the case of the output example from above it seems the messages are just there for debugging purposes / to know where the application currently is.
However: Switching the message mechanism of the applications to a more reasonable one is imho out of the scope for this patch.

For this patch I just used the existing mechanism for messages, but I could hide the additional output I introduced behind the `verbose` switch (just like the existing code does, though I would definitely prefer to print the messages in any case to be able to see what happened during execution in the log files).
Please let me know how we should proceed.

aschnell commented on client/mksubvolume.cc:
```diff
@@ -326,45 +396,83 @@ doit()
+						      MNT_ITER_BACKWARD);
+    if (subvolume_name.empty())
+	throw runtime_error("target is a dedicated mountpoint");
+    if (fstab_entry != NULL && strcmp(mnt_resolve_spec(mnt_fs_get_source(fstab_entry), cache),
```
At least mnt_resolve_spec () can return NULL causing a seg.fault. I can trigger this by writing a wrong UUID for / in fstab.

laenion replied:
Fixed, thanks.

aschnell commented on client/mksubvolume.cc:
```diff
@@ -326,45 +396,85 @@ doit()
 
-    // Execute all steps.
+    // Set up cache for UUID / LABEL resolution in mnt_table_find_source
+    libmnt_cache* cache = mnt_new_cache();
```
mnt_new_cache() could fail. More important is when cache gets freed, AFAIS a call to mnt_unref_cache() somethere later is missing.

laenion replied:
Error handling for `mnt_new_cache` is handled in `MntTable::set_cache` already. MntTable will now also dereference the cache automatically.

aschnell commented:
The function doit() increased a lot. Moving some code to helper functions could help a lot to understand the code.

laenion commented:
The consistency check code has been moved to a separate function now.

aschnell commented:
Unfortunately I have found another problem with the checks. A multiple device btrfs mounted by the device name in /etc/fstab. In /proc/mounts another device of the btrfs can be used and then the
check fails and reports "different device mounted on target".

aschnell commented on client/mksubvolume.cc:
```diff
@@ -259,6 +278,104 @@ find_filesystem(MntTable& mnt_table)
+    if (mounted_entry != NULL)
+    {
+	// Find last device in mtab to get the actual mount
+	mounted_entry = mtab_table.find_target(mnt_fs_get_target(expected_fs), MNT_ITER_BACKWARD);
```
Isn't the exact statement already used 20 lines above to set ```mount_entry```?

laenion replied:
Yes, it is - I forgot to remove that after several code refactorings...

aschnell commented on client/mksubvolume.cc:
```diff
@@ -259,6 +278,104 @@ find_filesystem(MntTable& mnt_table)
+do_consistency_checks(MntTable& mnt_table, libmnt_fs* fs, libmnt_fs* expected_fs)
+{
+    char *subvol_expected, *subvol_fstab, *subvol_real;
+    const char *dev_expected, *dev_fstab;
```
These variables should be declares as late as possible. Some are even only used in sub blocks.

laenion commented:
I'm not sure where the Changelog is stored, so here are the corresponding ticket numbers:

* https://bugzilla.suse.com/show_bug.cgi?id=1138725 (boo#1138725: Add Ignition to Tumbleweed)
* https://bugzilla.suse.com/show_bug.cgi?id=1126900 (boo#1126900: health-checker: grub2-editenv: error: cannot open `/boot/grub2/grubenv': Read-only file system.)
* https://github.com/openSUSE/snapper/issues/236 (don't throw in create_subvolume() if subvolume exists)

aschnell commented:
The changelog is in package/snapper.changes

aschnell commented:
looks good now

jpeaglesandkatz commented:
This problem is not fixed.

CachyOS latest updated 6.14 kernel.. KDE plasma 

" Creating config failed (creating btrfs subvolume .snapshots failed since it already exists)." 

I had to remove and reinstall snapper/btrfs suppor assistant because it was interfering with timeshift.

upon reinstall I cannot create a config in btrfs assistant because it complains the .btrfs subvolume already exists. Which seems obvious since they containt the snapshots from my previously installed snapper support.

How do I fix this issue now? I cannot use snapper/btrfs assistant at all.
